sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -416,6 +416,29 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         // these branches can be pruned away
         val (h, t) = branches.span(_._1 != TrueLiteral)
         CaseWhen( h :+ t.head, None)
+
+      case e @ CaseWhen(branches, _) =>
+        val newBranches = branches.foldLeft(List[(Expression, Expression)]()) {
```

Comment:
what about using `ArrayBuffer`? So we don't have to recreate the List at every iteration...